### PR TITLE
Use AM_CFLAGS instead of CFLAGS

### PR DESCRIPTION
**Pull request by tagoh**
As it is recommended way in general use case of automake

### CODE REVIEW

**Comment by tagoh** [comment]
That's the sort of FAQ. see https://www.gnu.org/software/automake/manual/html_node/Flag-Variables-Ordering.html

Plus, some distributions has own build options to enable some features in gcc say and it is passed through overriding them. the sort of this prevents to do so. and automake usually generates the build rules as is with AM_CFLAGS and CFLAGS. so I don't think it will breaks something even if someone overrides CFLAGS. you could use CFLAGS_FOR_BUILD and LDFLAGS_FOR_BUILD if you like. but you may want to put it into AM_CFLAGS and AM_LDFLAGS but not CFLAGS and LDFLAGS directly.

**Comment by Frenzie** [comment]
Your FAQ seems to confirm rather than assuage my fears.

Note that I am in no way related to this project. I'm just a user with a problem, namely that while setting CFLAGS and LDFLAGS for the cross-compiler, FriBiDi singularly fails to build because the preprocessor/compiler/linker quite understandably chokes on them in regard to these native helper executables.

**Comment by tagoh** [comment]
You had doubts on this change. that's fine. I'm willing to explain no matter who you are. that would helps someone too who might have similar concerns.

However you seem misunderstanding something here. you may be concerned that something important things in CFLAGS at this moment will be overridden by this change though, it won't happens, because autotools takes care of them separately. let's see. automake usually generates like:

COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)

and $COMPILE is used to build something. this simply explains both requirements for holding necessary options to build and things users want to add can be satisfied. so why people still see the problem around cross-compilation then? that's because those aren't used properly.
Basically whether users wants to do cross-compile it is detected in configure script. and all of build options must be essentially held in own variables instead of modifying CFLAGS/LDFLAGS directly, because that breaks the above framework and assumption in autotools. everything must be sane rather than bandaid on it.

That said, the buildsystem seems moving to meson here. any changes related to autotools might be not getting merged. so it may be good to try there.

**Comment by Frenzie** [comment]
Meson doesn't have this problem in the first place (or at least not since #88).

> that's because those aren't used properly.

That's berating, not explaining. ;-)

If I run ./configure, it quite explicitly says:

```
Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
  LT_SYS_LIBRARY_PATH
              User-defined run-time library search path.
  CPP         C preprocessor
  PKG_CONFIG  path to pkg-config utility
  PKG_CONFIG_PATH
              directories to add to pkg-config's search path
  PKG_CONFIG_LIBDIR
              path overriding pkg-config's built-in search path

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.
```

So if the strategy suggested by `configure` itself (!) is wrong, i.e., `CFLAGS=whatever LDFLAGS=lala ./configure --host=SOMEHOST`, what is the user supposed to do instead? (That's where an explanation would come in handy.)

Afaict you have `CC`, `CFLAGS`, `LDFLAGS`, etc. to customize the target if required, and the likes of `CC_FOR_BUILD`, `CFLAGS_FOR_BUILD`, `LDFLAGS_FOR_BUILD`, etc. to do the same for the build machine.

**Comment by tagoh** [comment]
The interactions for users and what developers follows (to make those interactions enabled) is different thing. I'm rather wondering how both requirements can be satisfied without any collisions by using same variables? if configure script and Makefile.am relies on CFLAGS/LDFLAGs etc, the above command you quoted will breaks the build easily, because what developers expects will be miseed by it then.  To avoid this situation, well, this is explained in the above url though, they said user variables such as CFLAGS should never be redefined as well as things in configure.

Running automake warns these days if they find such thing out in Makefile.am. this PR has been filed according to it.

BTW I'm not opposed to have the sort of *_FOR_BUILD. just saying they should go with the right way.

**Comment by Frenzie** [comment]
Please replace "the user" by "me" if that helps.

When I run `CFLAGS=whatever LDFLAGS=lala ./configure --host=SOMEHOST`, FriBiDi will subsequently fail to compile with current master if there's some target-specific flag in there unsupported by the native compiler. This is wrong. It should only fail when I, the user, stick an unsupported flag in `*_FOR_BUILD`.

My question above was quite simply whether AM_CFLAGS will do the right thing here, or if it will exacerbate the problem.

> BTW I'm not opposed to have the sort of *_FOR_BUILD. just saying they should go with the right way.

*_FOR_BUILD is the standard Autotools way. It's not some FriBiDi thing.

**Comment by Frenzie** [comment]
Reference: https://www.gnu.org/software/autoconf-archive/ax_prog_cc_for_build.html

**Comment by tagoh** [comment]
Well, are you sure where does those target-specific flags come from? and can it be fixed by calling AX_PROG_CC_FOR_BUILD and set to the certain build related variables such as LDFLAGS? if it works for non-cross build too, what's wrong with AM_* such as AM_CFLAGS instead of CFLAGS say? well, honestly speaking, I don't get what you are facing. in fact, I'm also an user of that macro and my autotools based project is capable to cross-compile. I don't see any problems with AM_*.

maybe good to try first before arguing. if it doesn't work, you may want to investigate where the problematic flags is coming from. that's what it's worth to try. again, using AM_* is the right way to go.

**Comment by Frenzie** [comment]
I wasn't arguing, I was asking. My apologies if that wasn't clear from the outset.

**Comment by Frenzie** [comment]
Unless of course you mean the stuff about *_FOR_BUILD. I am most definitely arguing about that.

**Comment by Frenzie** [comment]
Okay, tried it with AM_CFLAGS. Looks like my concerned question was quite on-point:

```
  CC       gen-unicode-version.o
gcc: error: unrecognized command line option ‘-mfpu=vfpv3-d16’
gcc: error: unrecognized command line option ‘-mthumb’; did you mean ‘-mtbm’?
gcc: error: unrecognized command line option ‘-mfloat-abi=softfp’
```

**Comment by tagoh** [comment]
Sure. hm, well, something wrong to configure. apparently you were using host gcc or native libraries/devel files including pkgconfig files to determine the gcc flags in Makefile. next steps would be to check where those flags are set in Makefile. probably good to check config.log too.

Anyway, that is another problem. preventing this PR due to that is just about to hide whole problems around it. that sounds not good to me.

**Comment by Frenzie** [comment]
You do realize that `gen.tab` is meant to be compiled by the native compiler? The way it currently achieves this is by strategically overriding `CC` with `CC_FOR_BUILD` and `CFLAGS` with `CFLAGS_FOR_BUILD`. The `*_FOR_BUILD` line of variables is all generated quite correctly. Everything is. AutoTools is awesome that way. *cough*Unlike CMake.*cough*

https://github.com/fribidi/fribidi/blob/aac3845708969b9c4bc2874edc4950123595b1b4/gen.tab/Makefile.am#L27-L28

I'm more than willing to believe there's a better way to use `*_FOR_BUILD`, but the simple fact is that this PR breaks any attempt at cross-compilation almost entirely (unless the native compiler happens not to choke on any target `CFLAGS`). Moreover, I've looked through the documentation and I'm pretty sure the `AM_*` line of variables is just as wrong for this file as I originally suspected, because those are what you should use for the target.

For reference, here's the warning message:
```
bin/Makefile.am: installing './depcomp'
lib/Headers.mk:25: warning: shell cat $(top_srcdir: non-POSIX variable name
lib/Headers.mk:25: (probably a GNU make extension)
doc/Makefile.am:16:   'lib/Headers.mk' included from here
doc/Makefile.am:21: warning: shell sed '$$d' $(top_srcdir: non-POSIX variable name
doc/Makefile.am:21: (probably a GNU make extension)
gen.tab/Makefile.am:29: warning: 'CFLAGS' is a user variable, you should not override it;
gen.tab/Makefile.am:29: use 'AM_CFLAGS' instead
gen.tab/Makefile.am:28: warning: 'LDFLAGS' is a user variable, you should not override it;
gen.tab/Makefile.am:28: use 'AM_LDFLAGS' instead
lib/Headers.mk:25: warning: shell cat $(top_srcdir: non-POSIX variable name
lib/Headers.mk:25: (probably a GNU make extension)
lib/Makefile.am:17:   'lib/Headers.mk' included from here
```

The warning says not to override the user's `CFLAGS`. They're not being overridden. The user's `CFLAGS` are being used perfectly fine while compiling the actual library, and the user's `CFLAGS_FOR_BUILD` are used while compiling these generators.

If this program is doing it the wrong way — well, frankly you still haven't told me how I would avoid the same mistake while writing my own Makefile.am. While I may be mistaken in my analysis, the PR as written seems to do nothing but follow the course of action suggested by the error message, while failing to identify the actual problem. Which is primarily that `CPPFLAGS`/`CFLAGS`/`LDFLAGS` should be overridden some other way to avoid triggering some noise (cf. [the situation two years ago](https://github.com/fribidi/fribidi/commit/ad1da2a0d79f4d362d03d2b21834be9c505c5b74) which probably doesn't trigger any warnings). Secondarily the cross-compilation support still looks a bit tacked on, which is of course partially related to the warning but it goes deeper than that.

This file should probably use `BUILD_EXEEXT` instead of `EXEEXT`, and to avoid the warning it could set various `gen_etc_CFLAGS/LDFLAGS/OBJECTS`. In lieu of that, please don't try to extend the scope of the `LDFLAGS` issue to `CFLAGS`. ;-)

**Comment by tagoh** [comment]
Hmm, I'm not trying to extend at all. I'm saying here to follow the right way as autotools are recommending as I mentioned first. there are nothing more than that.

Anyway, the default build system seems moving to meson. I don't think it's worth keeping discussion about autotools things.

**Comment by Frenzie** [comment]
Sorry, but I don't know how many different ways I can repeat my very first comment. The right way is `CFLAGS_FOR_BUILD` and `LDFLAGS_FOR_BUILD`. These are helper executables, meant to run on the machine on which you're compiling. They are *not* meant to take `CFLAGS` or `LDFLAGS`! What part of that is so hard to understand?

**Comment by tagoh** [comment]
You have said the answer yourself. yes, *_FOR_BUILD are supposed to work without directly setting CFLAGS/LDFLAGS and so on. but you want to take CFLAGS here. why don't you try to find the root cause out for your problem? you shold blame things where added the problematic options to CFLAGS and replace it with CFLAGS_FOR_BUILD but not the whole things.

Well, think of more, putting the kind of things in Makefile.am was probably wrong. I'm not carefully taking a look though, what you want to replace by *_FOR_BUILD may be coming from something in configure. google it and find out some projects supporting cross-compiling and see how they use *_FOR_BUILD there. you may have some clue how you can fix your problem.

Anyway, settings something directly to CFLAGS/LDFLAGS are wrong approach as autotools warns. if you still think it is wrong, you should try to contact autotools maintainers. discussing about it here is wrong place.

**Comment by Frenzie** [comment]
> but you want to take CFLAGS here

You. Do. Not.

`CFLAGS` and `LDFLAGS` are for the target, not the build system. These generators are *part of the build system*. They're like a Makefile or a shellscript, except they happen to be written in C.

Building on x86_64 for ARM? Using the ARM cross-compiler would be *wrong*. The generators wouldn't be able to run. Using the ARM flags would be *wrong*. The build compiler or linker may not support them.

You may *optionally* use `*_FOR_BUILD` (although I'm hard-pressed to think of a scenario where you'd even want to).

> why don't you try to find the root cause out for your problem?

Why do you think I know what these generators are? Back in September I took ten minutes to investigate so that I *actually understood what I was dealing with*.

You can't seem to get beyond the fact that it says `CFLAGS = `. This is indeed "wrong,", no one's ever disputed that. But good grief, quit acting like a computer script.

**Comment by tagoh** [comment]
> Building on x86_64 for ARM? Using the ARM cross-compiler would be wrong. The generators wouldn't be able to run. Using the ARM flags would be wrong. The build compiler or linker may not support them.

That's right. and that is the problem fribidi actually needs to be fixed/improved. it would be similar to make a cross-compiler-compiler. as you set up with configure to cross-compile fribidi, CFLAGS/LDFLAGS should be configured to cross-compile. using it to build a cross-compiler-compiler would be indeed wrong. that said, I still think replacing it with *_FOR_BUILD is not the right way. we'll miss an opportunity to add something to CFLAGS outside make or configure, which is my real problem I wanted to fix here.

Then what I was suggesting is that stop adding something to CFLAGS in configure even and have a separate *_CFLAGS and use them appropriately. that would be something meson more or less do.

Although I stopped using autotools to build fribidi and switched to meson. I'm happy with it. so I'm not that interested in to get this fixed anymore.

**Comment by Frenzie** [comment]
> we'll miss an opportunity to add something to CFLAGS outside make or configure, which is my real problem I wanted to fix here. […] switched to meson. I'm happy with it.

Yet Meson is actively hostile to your proclaimed goal. Meson actively sanity checks CFLAGS for native and cross with the env and it's horribly broken for exactly the same reason as this PR.

https://github.com/mesonbuild/meson/pull/5128
https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/meson.eclass#n123

**Comment by Frenzie** [comment]
NB If Meson actually did it right (without bugs/leakage) it might be okay, but that's besides the general point that it opposes what you say you want.

**Comment by Frenzie** [comment]
Anyway, that finally makes sense, although I can't fathom for the life of me why even now you seem to be implying AM_CFLAGS is the "right way" when it's obviously not.
